File: app/crud/vault.py
import uuid
from datetime import datetime

from sqlalchemy.orm import Session

from .. import models, schemas


def get_vault(db: Session, user_id: str, vault_id: str):
    return (
        db.query(models.Vault)
        .filter(models.Vault.id == vault_id)
        .filter(models.Vault.user_id == user_id)
        .first()
    )


def get_vaults_by_user_id(db: Session, user_id: str, skip: int = 0, limit: int = 100):
    return (
        db.query(models.Vault)
        .filter(models.Vault.user_id == user_id)
        .offset(skip)
        .limit(limit)
        .all()
    )


def create_vault(db: Session, vault: schemas.VaultCreate, user_id: str):
    vault_id = uuid.uuid4()
    db_vault = models.Vault(
        name=vault.name,
        tags=vault.tags,
        user_id=user_id,
        id=vault_id,
        is_default=vault.is_default,
        created_at=datetime.utcnow(),
    )
    db.add(db_vault)
    return db_vault


def update_vault(db: Session, db_vault: schemas.Vault, vault: schemas.VaultUpdate):
    for field, value in vault.dict(exclude_unset=True).items():
        setattr(db_vault, field, value)
    return db_vault


def delete_vault(db: Session, db_vault: schemas.Vault):
    db.delete(db_vault)
    return db_vault

File: go_app/server/crud/vault.go
package crud

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"

	"github.com/google/uuid"
	"terafill.com/m/server/schemas"
)

func GetVault(ctx context.Context, db *sql.DB, userId string, vaultId string) (*schemas.Vault, error) {
	var vault schemas.Vault

	row := db.QueryRowContext(ctx, "SELECT id, name, description, user_id, created_at, tags, is_default FROM vaults WHERE user_id = ? and vault_id = ?", userId, vaultId)
	if err := row.Scan(&vault.ID, &vault.Name, &vault.Description, &vault.UserID, &vault.CreatedAt, &vault.Tags, &vault.IsDefault); err != nil {
		if err == sql.ErrNoRows {
			return &vault, fmt.Errorf("GetVault %v: no such vault found", vaultId)
		}
		return &vault, fmt.Errorf("GetVault %v: %v", userId, err)
	}
	return &vault, nil
}

func GetVaultsByUserId(ctx context.Context, db *sql.DB, userId string) ([]schemas.Vault, error) {
	var vaults []schemas.Vault

	rows, err := db.QueryContext(ctx, "SELECT id, name, description, user_id, created_at, tags, is_default FROM vaults WHERE user_id = ?;", userId)
	if err != nil {
		return nil, fmt.Errorf("GetVaults %q: %v", userId, err)
	}

	defer rows.Close()

	for rows.Next() {
		var vault schemas.Vault
		var tags interface{}
		if err := rows.Scan(&vault.ID, &vault.Name, &vault.Description, &vault.UserID, &vault.CreatedAt, &tags, &vault.IsDefault); err != nil {
			return vaults, fmt.Errorf("GetVaults %v: %v", userId, err)
		}
		var tagsBytes = tags.([]byte)
		json.Unmarshal(tagsBytes, &vault.Tags)
		vaults = append(vaults, vault)
	}

	return vaults, nil
}

func CreateVault(ctx context.Context, db *sql.DB, vault *schemas.Vault) error {
	vault.ID = uuid.Must(uuid.NewRandom()).String()
	tags, err := json.Marshal(vault.Tags)
	_, err = db.ExecContext(ctx, "INSERT INTO vaults (id, name, description, user_id, tags, is_default) VALUES (?, ?, ?, ?, ?, ?)",
		vault.ID,
		vault.Name,
		vault.Description,
		vault.UserID,
		string(tags),
		vault.IsDefault,
	)
	if err != nil {
		return fmt.Errorf("CreateVault: %v", err)
	}
	return nil
}

func UpdateVault(ctx context.Context, db *sql.DB, vault *schemas.Vault) error {
	tags, err := json.Marshal(vault.Tags)
	_, err = db.ExecContext(ctx, "UPDATE vaults set name=?, description=?, user_id=?, tags=?, is_default=? where id=?;",
		vault.Name,
		vault.Description,
		vault.UserID,
		string(tags),
		vault.IsDefault,
		vault.ID,
	)
	if err != nil {
		return fmt.Errorf("UpdateVault: %v", err)
	}
	return nil
}

func DeleteVault(ctx context.Context, db *sql.DB, vaultId string) error {
	_, err := db.ExecContext(ctx, "DELETE from vaults where id=?;",
		vaultId,
	)
	if err != nil {
		return fmt.Errorf("DeleteVault: %v", err)
	}
	return nil
}
